Format code and docs with black+blacken-docs+pre-commit

sloria opened the pull request:
* Format code with [black](https://github.com/ambv/black)
* Format doc code blocks with [blacken-docs](https://github.com/asottile/blacken-docs/). This uncovered a number of syntax issues in the docs.
* Run both with [pre-commit](https://pre-commit.com/)

sloria committed:
Format code and docs with black+blacken-docs+pre-commit

sloria commented:
@asottile If you're feeling generous, feel free to give this a drive-by =).

asottile commented on docs/conf.py:
```diff
@@ -6,51 +6,51 @@
-sys.path.insert(0, os.path.abspath('..'))
-import webargs
+sys.path.insert(0, os.path.abspath(".."))
+import webargs  # flake8: noqa
```
[fun story about `# flake8: noqa`](https://gitlab.com/pycqa/flake8/merge_requests/219)

I _believe_ you want `# noqa` here (or `# flake8: noqa` on a line by itself if you actually don't want to lint this file)

sloria replied:
Ah, TIL. Thanks.

asottile commented on setup.cfg:
```diff
@@ -2,6 +2,8 @@
-max-line-length = 100
-exclude = .git,.ropeproject,.tox,build,.direnv,__pycache__
+ignore = E203, E266, E501, W503, E302
+max-line-length = 80
```
black I believe uses `88`, though I see you've ignored `E501` so I guess this line does nothing! :D

lafrech replied:
Is it possible to use explicit names in there?

    -ignore = E203, E266, E501, W503, E302
    +ignore = rule-number-one, another-rule

asottile replied:
unfortunately not -- pylint has this but flake8 does not

sloria replied:
> black I believe uses 88, though I see you've ignored E501 so I guess this line does nothing! :D

This is from the [recommended flake8 config from black](https://github.com/ambv/black/blob/master/.flake8). E501 is ignored, but Bugbear's B901 is enabled, which will warn when the max-line-length exceeds 10% of `max-line-length`.

asottile replied:
are you using `flake8-bugbear`? I don't see it installed

sloria replied:
Ah, woops. Forgot to add that to dev-requirements.txt. Will fix this up after work today.

asottile commented on tasks.py:
```diff
@@ -5,79 +5,95 @@
+
+@task
+def precommit(ctx):
+    ctx.run("pre-commit run --all-files", echo=True)
```
`--show-diff-on-failure` is usually a good option for CI's sake (helps drive-by contributors that might not have set up the linting locally)

asottile reviewed (commented):
Looks good!  Just a few small things I noticed :)

lafrech commented:
> Black can be installed by running pip install black. It requires Python 3.6.0+ to run but you can reformat Python 2 code with it, too.

My local environment is Python 3.5. Will this make `pre-commit` fail?

asottile commented:
> My local environment is Python 3.5. Will this make `pre-commit` fail?

asottile commented:
> My local environment is Python 3.5. Will this make `pre-commit` fail?

Not necessarily, the hooks here are configured with `language_version: python3.6` so pre-commit will attempt to instantiate hook environments using that interpreter -- if there's no python3.6 then yes it will fail hook installation and tell you so.

sloria commented:
Addressed @asottile 's comments and also added some notes in CONTRIBUTING.rst about needing the `python3.6` interpreter for `pre-commit` to work properly. If you don't have it installed, you can still run tests with 

```
inv test --no-syntax
```

lafrech commented:
OK. This means that all contributors on Python 3.5 won't be able to use `pre-commit` anymore.

There's a trade-off, here. This commit adds more features to `pre-commit` but prevents part of the contributors to use it at all. I don't have stats on who uses 3.5 or 3.6. Using Debian Stable, I'll probably be using 3.5 for a while.

No big deal. If I can't run it locally, Travis will complain and I'll use my special ability (`git push -f`) to fix my mess.

sloria commented:
@lafrech You can use Python 3.5 in your virtualenv and still use pre-commit as long as you have the `python3.6` interpreter globally.

With pyenv:

```
pyenv install 3.6.5
pyenv global system 3.6.5
```

sloria committed:
Use flake8-bugbear; use async def coroutines
